doc/subsystems/codegen/codegen.rst

```diff
@@ -0,0 +1,409 @@
+..
+    Copyright (c) 2004-2015 Ned Batchelder
+    SPDX-License-Identifier: MIT
+    Copyright (c) 2018 Bobby Noelte
+    SPDX-License-Identifier: Apache-2.0
+
+.. _codegen:
+
+Inline Code Generation
+######################
+
+For some repetitive or parameterized coding tasks, it's convenient to
+use a code generating tool to build C code fragments, instead of writing
+(or editing) that source code by hand. Such a tool can also access CMake build
+parameters and device tree information to generate source code automatically
+tailored and tuned to a specific project configuration.
+
+The Zephyr project supports a code generating tool that processes embedded
+Python "snippets" inlined in your source files. It can be used, for example,
+to generate source code that creates and fills data structures, adapts
+programming logic, creates configuration-specific code fragments, and more.
+
+.. contents::
+   :depth: 2
+   :local:
+   :backlinks: top
+
+Description
+***********
+
+Python snippets that are inlined in a source file are used as code generators.
+The tool to scan the source file for the Python snippets and process them is
+Codegen. Codegen and part of this documentation is based on
+`Cog <https://nedbatchelder.com/code/cog/index.html>`_ from Ned Batchelder.
+
+The processing of source files is controlled by the CMake extension functions:
+zephyr_sources_codegen(..) or zephyr_library_sources_codegen(..). The generated
+source files are added to the Zephyr sources. During build the source files are
+processed by Codegen and the generated source files are written to the CMake
+binary directory.
+
+The inlined Python snippets can contain any Python code, they are regular
+Python scripts. All Python snippets in a source file and all Python snippets of
+included template files are treated as a python script with a common set of
+global Python variables. Global data created in one snippet can be used in
+another snippet that is processed later on. This feature is e.g. used to
+customize included template files.
+
+An inlined Python snippet can always access the codegen module. The codegen
+module encapsulates and provides all the functions to retrieve information
+(options, device tree properties, CMake variables, config properties) and to
+put out the generated code.
+
+Codegen transforms files in a very simple way: it finds chunks of Python code
+embedded in them, executes the Python code, and inserts its output back into
```

Comment:
To be sure here, when you say it "executes the Python code and inserts it output into the original file" should I take that literally?  If my Python code says ``print("hello");`` will ``hello`` get put into the generated output?  If not into the generated output, where would the ``hello`` get written or does it go to ``/dev/null``?